### ISSUE TITLE
torch.pow has incorrect backward result when the base is real negative and exponent is complex

### ISSUE THREAD

**Issue by iclementine** [pytorch/pytorch#89903]
### 🐛 Describe the bug

`torch.pow(x, y)` has incorrect backward result when the base (`x`)  is real negative and exponent(`y`) is complex. 

Because the backward propagation to the `y` uses `log(x)` in its computation. But when `x` is real and negative, the result is nan. But in the case that any one in `x` and `y` is complex, the pow should be considered a complex function.

```
pow(x, y) = exp(log(x) * y)

dpow(x, y) / dx = y * pow(x, y - 1)
dpow(x, y) / dy = pow(x, y) * log(x)
```

According to torch's convention of complex gradient. the backward of x and y are
https://github.com/pytorch/pytorch/blob/4cb6bbbe27162c7b0835879131991d2155329718/torch/csrc/autograd/FunctionsManual.cpp#L441-L459

```
g_x = (grad.conj() * y * pow(x, y - 1)).conj()
g_y =  (grad.conj() * pow(x, y) * log(x)).conj()
```
respectively. but the log(x) should be considered function on complex plane. So I propose to fix it with the following change:

```
g_x = (grad.conj() * y * pow(x, y - 1)).conj()
g_y =  (grad.conj() * pow(x, y) * log(x.astype(y.dtype))).conj()
```

That is, to change L457, `self.log()` to `self.to(exponent.dtype()).log()`

## Code to reproduce

```
import jax
import numpy as np
import torch
np.random.seed(42)


x = np.random.randn(3)
y = np.random.randn(3) + 1j * np.random.randn(3)
g = np.random.randn(3) + 1j * np.random.randn(3)

print("x: ", x)
print("y: ", y)
print("============================")

device = torch.device("cpu:0")
x1 = torch.tensor(x, device=device, requires_grad=True)
y1 = torch.tensor(y, device=device, requires_grad=True)
g1 = torch.tensor(g, device=device)
o1 = torch.pow(x1, y1)
o1.backward(g1)
print("o1: \n", o1.detach().cpu().numpy())
print("x1g: \n", x1.grad.detach().cpu().numpy())
print("y1g: \n", y1.grad.detach().cpu().numpy())


print("============================")
o2, f_vjp = jax.vjp(jax.numpy.power, x, y)
x2g, y2g = f_vjp(g.conj()) # note that Jax use a different convention for complex gradient
print("o2: \n", np.asarray(o2))
print("x2g: \n", np.asarray(x2g))
print("y2g: \n", np.asarray(y2g))
```

the results
```
x:  [ 0.49671415 -0.1382643   0.64768854]
y:  [ 1.52302986+1.57921282j -0.23415337+0.76743473j -0.23413696-0.46947439j]
============================
o1:
 [ 0.15470474-0.30778461j -0.09002977-0.11059353j  1.08411083+0.22418126j]
x1g:
 [ 0.67894989 -0.24261711  1.60209266]
y1g:
 [-0.00662259-0.143044j           nan       +nanj  0.38726098+0.76687764j]
============================
o2:
 [ 0.15470475-0.30778462j -0.09002977-0.11059352j  1.0841109 +0.22418123j]
x2g:
 [ 0.6789499  -0.24261728  1.6020925 ]
y2g:
 [-0.00662259+0.14304401j -0.12107516+1.0352328j   0.3872609 -0.7668776j ]
```

Thought I found this on torch 1.12, the related code is still the same in current version.

### Versions

PyTorch version: 1.12.0a0+git369d9f4
Is debug build: True
CUDA used to build PyTorch: 11.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (GCC) 8.2.0
CMake version: version 3.22.3
Libc version: glibc-2.27

Python version: 3.9.2 (default, Feb 20 2021, 20:57:50)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-4.14.0_1-0-0-41-x86_64-with-glibc2.27

....

Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.1.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.6
[pip3] torch==1.12.0a0+git369d9f4
[pip3] torchvision==0.12.0
[conda] Could not collect

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @anjali411 @dylanbespalko @mruberry

**Comment by lezcano** [comment]
The idea is somehow correct, but the issue is a bit more subtle, What if it's the other way around? If `base` is complex and the exponent is real, I believe your change would give incorrect gradients.

I think the correct fix would go through:
- Compute the "max" of their two dtypes according to the order we have in PyTorch (there's a function for this) and cast both arguments to that dtype. `handle_r_to_c` already projects onto the real numbers when needed, so everything should work as expected in the real case. 
- Add some samples testing these cases (e.g. complex & float, float & complex, float & double) to the relevant OpInfo to make sure everything's working as expected.

We would accept a PR implementing this.

**Comment by iclementine** [comment]
> What if it's the other way around? If `base` is complex and the exponent is real, I believe your change would give incorrect gradients.

Yes, you are right. The base should be cast to the promoted type of the base and the exponent.

I would work on implementing it. Thank you for your reply.

**Comment by lezcano** [comment]
the base and the exponent should be both casted to the largest of the two types, yes. And sure, feel free to put up a PR and tag me as a reviewer in it.
